### ISSUE TITLE
Add support for PEP-631 (PEP 621)

### ISSUE THREAD

**Issue by jancespivo** [jazzband/pip-tools#1552]
#### What's the problem this feature will solve?

PEP-631 describes dependencies can be declared in pyproject.toml under `[project.dependencies]` section (see https://www.python.org/dev/peps/pep-0631/#dependencies):
```
[project]
dependencies = [
...
]
```

It is tool-independent way how to describe dependencies instead of tool-specific ones like `[tool.poetry.dependencies]`

#### Describe the solution you'd like

Add support for https://www.python.org/dev/peps/pep-0631.

**Comment by AndydeCleyre** [comment]
AFAIU this is already supported in terms of using `pyproject.toml` as a *source*, so this request is specifically about using `pyproject.toml` as a *destination*, right?

But then, the requirements you'd put there are IMO more likely and more reasonably equivalent to the "loose" requirements defined in your `*.in` files, rather than the compiled result. I do have a zsh function `pypc` which populates a `pyproject.toml` with the `*.in` entries, recently demoed at #1495.

**Comment by AndydeCleyre** [comment]
So this is a request to support the new spec in the file as a *source*, then right?

**Comment by AndydeCleyre** [comment]
Some notes:

- we currently use this to intake setup files including `pyproject.toml`:

```python
from pep517 import meta
dist = meta.load(os.path.dirname(os.path.abspath(src_file)))
```

- discussion at pypa/pip#1884 ([comment](https://github.com/pypa/pip/issues/1884#issuecomment-941137347)):

> No-one has yet looked at PEP 621 support in pip - apart from the obvious "no-one has been interested enough" reason, I think it's a matter of waiting until enough projects publish PEP 621 compatible metadata (and specifically *dependency* metadata) to make it worthwhile. Personally, I think that when setuptools starts writing PEP 621 data is probably the point at which I'd see it as worthwhile for pip (there will be a lot of projects that won't be able to generate static dependencies, even if setuptools supports them, but at least having the *possibility* seems like a reasonable starting point). Of course, you might feel differently about whether we should wait on a particular backend 

- See also: pypa/setuptools#1688

**Comment by jancespivo** [comment]
> So this is a request to support the new spec in the file as a _source_, then right?

Yes, exactly.

**Comment by jancespivo** [comment]
Original PEP-631 has been accepted and was merged into [PEP 621](https://peps.python.org/pep-0621). It is the same however it might add some context:
https://peps.python.org/pep-0621/#dependencies-optional-dependencies

**Comment by AndydeCleyre** [comment]
Can we get a good example, with expected vs actual output?

**Comment by jancespivo** [comment]
> Can we get a good example, with expected vs actual output?

Currently there is no output when dependencies are declared in pyproject.toml under [project.dependencies] section. The section is ignored by pip-tools completely.

Expected output would be the same as if dependencies was declared under [tool.poetry.dependencies] section (which is already implemented in pip-tools).

**Comment by AndydeCleyre** [comment]
@jancespivo 

I'm asking for a good example, because my current test case is a bad example, in that it seems to work:

```console
$ git clone https://github.com/AndydeCleyre/ptrender
$ cd ptrender
$ python3 -m venv venv
$ . ./venv/bin/activate
$ pip install -U pip-tools
$ cat pyproject.toml
```
```toml
[build-system]
requires = ["flit_core >=3.2,<4"]
build-backend = "flit_core.buildapi"

[project]
name = "ptrender"
authors = [{name = "Andy Kluger"}]
readme = "README.rst"
license = {file = "LICENSE"}
keywords = ["pyratemp", "template", "render", "renderer"]
requires-python = ">=3.6"
dynamic = ["version", "description"]
dependencies = ["plumbum", "pyratemp"]
classifiers = [
    "Topic :: Text Processing :: Markup",
    "Topic :: Utilities",
]

[project.urls]
Home = "https://github.com/andydecleyre/ptrender"

[project.optional-dependencies]
yaml = ["strictyaml"]
dev = ["flit", "ipython"]

[project.scripts]
ptrender = "ptrender.ptrender:TemplateRenderer"
vwrite = "ptrender.ptrender:VWriter"
```
```console
$ pip-compile --annotation-style=line --no-header -o - pyproject.toml
```
```ini
plumbum==1.7.2            # via ptrender (pyproject.toml)
pyratemp==0.3.2           # via ptrender (pyproject.toml)
```
```console
$ pip-compile --annotation-style=line --no-header -o - --extra yaml pyproject.toml
```
```ini
plumbum==1.7.2            # via ptrender (pyproject.toml)
pyratemp==0.3.2           # via ptrender (pyproject.toml)
python-dateutil==2.8.2    # via strictyaml
six==1.16.0               # via python-dateutil
strictyaml==1.6.1         # via ptrender (pyproject.toml)
```
```console
$ pip-compile --annotation-style=line --no-header -o - --extra dev pyproject.toml
```
```ini
asttokens==2.0.5          # via stack-data
backcall==0.2.0           # via ipython
certifi==2022.5.18.1      # via requests
charset-normalizer==2.0.12  # via requests
decorator==5.1.1          # via ipython
docutils==0.18.1          # via flit
executing==0.8.3          # via stack-data
flit==3.7.1               # via ptrender (pyproject.toml)
flit-core==3.7.1          # via flit
idna==3.3                 # via requests
ipython==8.3.0            # via ptrender (pyproject.toml)
jedi==0.18.1              # via ipython
matplotlib-inline==0.1.3  # via ipython
parso==0.8.3              # via jedi
pexpect==4.8.0            # via ipython
pickleshare==0.7.5        # via ipython
plumbum==1.7.2            # via ptrender (pyproject.toml)
prompt-toolkit==3.0.29    # via ipython
ptyprocess==0.7.0         # via pexpect
pure-eval==0.2.2          # via stack-data
pygments==2.12.0          # via ipython
pyratemp==0.3.2           # via ptrender (pyproject.toml)
requests==2.27.1          # via flit
six==1.16.0               # via asttokens
stack-data==0.2.0         # via ipython
tomli==2.0.1              # via flit
tomli-w==1.0.0            # via flit
traitlets==5.2.1.post0    # via ipython, matplotlib-inline
urllib3==1.26.9           # via requests
wcwidth==0.2.5            # via prompt-toolkit

# The following packages are considered to be unsafe in a requirements file:
# setuptools
```
```console
$ pip-compile --annotation-style=line --no-header -o - --extra dev,yaml pyproject.toml
```
```ini
asttokens==2.0.5          # via stack-data
backcall==0.2.0           # via ipython
certifi==2022.5.18.1      # via requests
charset-normalizer==2.0.12  # via requests
decorator==5.1.1          # via ipython
docutils==0.18.1          # via flit
executing==0.8.3          # via stack-data
flit==3.7.1               # via ptrender (pyproject.toml)
flit-core==3.7.1          # via flit
idna==3.3                 # via requests
ipython==8.3.0            # via ptrender (pyproject.toml)
jedi==0.18.1              # via ipython
matplotlib-inline==0.1.3  # via ipython
parso==0.8.3              # via jedi
pexpect==4.8.0            # via ipython
pickleshare==0.7.5        # via ipython
plumbum==1.7.2            # via ptrender (pyproject.toml)
prompt-toolkit==3.0.29    # via ipython
ptyprocess==0.7.0         # via pexpect
pure-eval==0.2.2          # via stack-data
pygments==2.12.0          # via ipython
pyratemp==0.3.2           # via ptrender (pyproject.toml)
python-dateutil==2.8.2    # via strictyaml
requests==2.27.1          # via flit
six==1.16.0               # via python-dateutil
stack-data==0.2.0         # via ipython
strictyaml==1.6.1         # via ptrender (pyproject.toml)
tomli==2.0.1              # via flit
tomli-w==1.0.0            # via flit
traitlets==5.2.1.post0    # via ipython, matplotlib-inline
urllib3==1.26.9           # via requests
wcwidth==0.2.5            # via prompt-toolkit

# The following packages are considered to be unsafe in a requirements file:
# setuptools
```

Is this because the flit backend implements the standard, and poetry doesn't? If so, is it just a matter implementing it in poetry?

**Comment by thisstillwill** [comment]
Any updates on this? With `pyproject.toml` I've been able to move most of my project configs for `pytest`, `isort`, etc. into a single file. It'd be great to do the same for the dependencies I have in `requirements.in` right now

**Comment by ssbarnea** [comment]
I migrated like 10+ projects to PEP-621 and using pip-tools with it, not sure what else is needed on that aspect. Maybe the only missing bit is that pip-tools does not include pyproject.toml among the implicit input files. I was planning to fix that, but once does is done we can fully close this.

Shortly, already done, is just a small tuning that could make ease easier, nothing more.

**Comment by lig** [comment]
pip compile crashes on this

```toml
[project.dependencies]
pydantic = { git = 'https://github.com/pydantic/pydantic.git' }

```

```shell
❯ pip-compile -q --resolver backtracking --extra all -o requirements/pyproject.tx
t pyproject.toml
Backend subprocess exited when trying to invoke prepare_metadata_for_build_wheel
Failed to parse /home/lig/workspace/pydantic/pydantic-extra-types/pyproject.toml
```

So the PEP isn't fully supported yet

**Comment by AndydeCleyre** [comment]
@lig

Can you provide a complete `pyproject.toml` to test against?

**Comment by lig** [comment]
Here you go https://github.com/pydantic/pydantic-extra-types/blob/lig/depend-on-pydantic-git/pyproject.toml

**Comment by AndydeCleyre** [comment]
> Here you go https://github.com/pydantic/pydantic-extra-types/blob/lig/depend-on-pydantic-git/pyproject.toml

This doesn't seem specific to pip-tools, as pip complains that

> Field `project.dependencies` must be an array

```console
$ python -V
Python 3.10.6
$ pip-compile --version
pip-compile, version 6.14.0
$ pip -V
pip 23.1.2 from /home/andy/.local/share/venvs/280f99c42fd5b344dc3c7476f43b07e0/venv/lib/python3.10/site-packages/pip (python 3.10)
$ git checkout lig/depend-on-pydantic-git
$ pip install --dry-run '.[all]'
Processing /home/andy/Code/pydantic-extra-types
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [20 lines of output]
      Traceback (most recent call last):
        File "/home/andy/.local/share/venvs/280f99c42fd5b344dc3c7476f43b07e0/venv/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/home/andy/.local/share/venvs/280f99c42fd5b344dc3c7476f43b07e0/venv/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/home/andy/.local/share/venvs/280f99c42fd5b344dc3c7476f43b07e0/venv/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 152, in prepare_metadata_for_build_wheel
          whl_basename = backend.build_wheel(metadata_directory, config_settings)
        File "/tmp/pip-build-env-nw2dzyxm/overlay/lib/python3.10/site-packages/hatchling/build.py", line 56, in build_wheel
          return os.path.basename(next(builder.build(wheel_directory, ['standard'])))
        File "/tmp/pip-build-env-nw2dzyxm/overlay/lib/python3.10/site-packages/hatchling/builders/plugin/interface.py", line 93, in build
          self.metadata.validate_fields()
        File "/tmp/pip-build-env-nw2dzyxm/overlay/lib/python3.10/site-packages/hatchling/metadata/core.py", line 244, in validate_fields
          self.core.validate_fields()
        File "/tmp/pip-build-env-nw2dzyxm/overlay/lib/python3.10/site-packages/hatchling/metadata/core.py", line 1325, in validate_fields
          getattr(self, attribute)
        File "/tmp/pip-build-env-nw2dzyxm/overlay/lib/python3.10/site-packages/hatchling/metadata/core.py", line 1200, in dependencies
          self._dependencies = list(self.dependencies_complex)
        File "/tmp/pip-build-env-nw2dzyxm/overlay/lib/python3.10/site-packages/hatchling/metadata/core.py", line 1166, in dependencies_complex
          raise TypeError(message)
      TypeError: Field `project.dependencies` must be an array
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

PEP 631 says it's been superseded by PEP 621. Neither PEP 631 nor PEP 621 seem to include that non-list syntax, but 621 notes that the updated spec is [here](https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#declaring-project-metadata). That document also doesn't mention the non-list syntax, and says:

> For `dependencies`, it is a key whose value is an array of strings. Each string represents a dependency of the project and MUST be formatted as a valid [PEP 508](https://peps.python.org/pep-0508/) string.

I don't see anything like that `{ git = ...` syntax at PEP 508 either.

**Comment by q0w** [comment]
It should be 
```toml
dependencies = [ "pydantic @ git+https://github.com/pydantic/pydantic.git"]
```

**Comment by astrojuanlu** [comment]
> I don't see anything like that `{ git = ...` syntax at PEP 508 either.

Yeah, looks like a Hatch/hatchling specific thing.

**Comment by AndydeCleyre** [comment]
Can anyone provide a [PEP 621](https://peps.python.org/pep-0621/) sample that pip-tools doesn't currently handle?

**Comment by AndydeCleyre** [comment]
@jancespivo @thisstillwill 

This issue doesn't yet have enough information to do anything with, namely a sample input and output that is not working as desired. If no one can supply that, I'm inclined to close this as in-actionable. Can you offer any more info?

**Comment by jancespivo** [comment]
@AndydeCleyre I've checked it and it works already, probably fixed in different issue.

pyproject.toml:
```toml
name = "some-package"
version = "0.0.0.dev"
requires-python = ">=3.8"

dependencies = [
    "kubernetes",
    "pyyaml",
    "pydantic",
]
```

```shell
pip-compile
```

Output:
```
#
# This file is autogenerated by pip-compile with Python 3.11
# by the following command:
#
#    pip-compile
#
annotated-types==0.6.0
    # via pydantic
cachetools==5.3.2
    # via google-auth
certifi==2023.7.22
    # via
    #   kubernetes
    #   requests
charset-normalizer==3.3.2
    # via requests
...
```    

From my point of view it is ok to close this issue. Don't know what others have to say.

Thanks!
